tpm2/seal.go

```diff
@@ -0,0 +1,445 @@
+	if c == nil {
+		return
+	}
+
+	if err == nil {
+		return
+	}
+
+	index, err := tpm2.CreateNVIndexResourceContextFromPublic(c.pub)
+	if err != nil {
+		return
+	}
+	c.tpm.NVUndefineSpace(c.tpm.OwnerHandleContext(), index, c.session)
+}
+
+// makeKeyDataPolicy creates the policy data required to seal a key with makeKeyDataWithPolicy
+// and creates a PCR policy counter if required.
+func makeKeyDataPolicy(tpm *tpm2.TPMContext, pcrPolicyCounterHandle tpm2.Handle, authKey secboot.AuxiliaryKey,
+	session tpm2.SessionContext) (data *keyDataPolicyParams, pcrPolicyCounter *createdPcrPolicyCounter,
+	authKeyOut secboot.AuxiliaryKey, err error) {
+	// Create an auth key.
+	if authKey == nil {
+		authKey = make(secboot.AuxiliaryKey, 32)
+		if _, err := rand.Read(authKey); err != nil {
+			return nil, nil, nil, xerrors.Errorf("cannot create key for signing dynamic authorization policies: %w", err)
+		}
+	}
+	authPublicKey, err := newPolicyAuthPublicKey(authKey)
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot derive public area of key for signing dynamic authorization policies: %w", err)
+	}
+
+	// Create PCR policy counter, if requested.
+	var pcrPolicyCount uint64
+	if pcrPolicyCounterHandle != tpm2.HandleNull {
+		if tpm == nil {
+			return nil, nil, nil, errors.New("cannot create a PCR policy counter without a TPM connection")
+		}
+
+		var pub *tpm2.NVPublic
+		pub, pcrPolicyCount, err = createPcrPolicyCounter(tpm, pcrPolicyCounterHandle, authPublicKey, session)
+		switch {
+		case tpm2.IsTPMError(err, tpm2.ErrorNVDefined, tpm2.CommandNVDefineSpace):
+			return nil, nil, nil, TPMResourceExistsError{pcrPolicyCounterHandle}
+		case isAuthFailError(err, tpm2.CommandNVDefineSpace, 1):
+			return nil, nil, nil, AuthFailError{tpm2.HandleOwner}
+		case err != nil:
+			return nil, nil, nil, xerrors.Errorf("cannot create new PCR policy counter: %w", err)
+		}
+
+		pcrPolicyCounter = &createdPcrPolicyCounter{
+			tpm:     tpm,
+			session: session,
+			pub:     pub}
+
+		defer func() { pcrPolicyCounter.undefineOnError(err) }()
+	}
+
+	nameAlg := tpm2.HashAlgorithmSHA256
+
+	// Create the initial policy data
+	policyData, authPolicy, err := newKeyDataPolicy(nameAlg, authPublicKey, pcrPolicyCounter.Pub(), pcrPolicyCount)
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot create initial policy data: %w", err)
+	}
+
+	return &keyDataPolicyParams{
+		Alg:    nameAlg,
+		Data:   policyData,
+		Digest: authPolicy}, pcrPolicyCounter, authKey, nil
+}
+
+type keyDataParams struct {
+	PCRPolicyCounterHandle tpm2.Handle
+	PCRProfile             *PCRProtectionProfile
+}
+
+// makeNewKeyData protects the supplied keys using the supplied keySealer and
+// parameters. If required, a PCR policy counter is created. The returned key
+// will have an initial PCR policy as specified via the supplied parameters.
+func makeNewKeyData(tpm *tpm2.TPMContext, key secboot.DiskUnlockKey, authKey secboot.AuxiliaryKey, params *keyDataParams,
```

Comment:
why there's a New in this method name? it doesn't carry a lot of clear information